torch/testing/_internal/common_distributed.py

```diff
@@ -102,21 +102,22 @@ class DistTestCases:
     backend_feature["plugin"] = set()
 
 
+def exit_if_lt_x_gpu(x):
+    if torch.cuda.device_count() < x:
+        sys.exit(TEST_SKIPS[f"multi-gpu-{x}"].exit_code)
+
+
 def skip_if_no_gpu(func):
     """Skips if the world size exceeds the number of GPUs, ensuring that if the
     test is run, each rank has its own GPU via ``torch.cuda.device(rank)``."""
 
     @wraps(func)
     def wrapper(*args, **kwargs):
-        if not torch.cuda.is_available():
-            sys.exit(TEST_SKIPS["no_cuda"].exit_code)
-        world_size = int(os.environ["WORLD_SIZE"])
-        if torch.cuda.device_count() < world_size:
-            sys.exit(TEST_SKIPS[f"multi-gpu-{world_size}"].exit_code)
+        exit_if_lt_x_gpu(int(os.environ["WORLD_SIZE"]))
```

Comment:
can you comment on the state of this code? is this still 'ok' to use exit or is this just an area that needs further work and not handled in this PR?

Reply:
I tried to explain this in https://github.com/pytorch/pytorch/pull/122397#issuecomment-2014941519

In short: It needs further work and might not even be possible.

Question here is when exactly `$WORLD_SIZE` is set. I.e. is this available when the decorator is executed or only when the function is executed, i.e. after the test setup (likely involving forking). In the latter case this cannot be helped.

Similar below: Dependency on `self.world_size` and we don't have `self` available in the decorator. And especially as this property can be set/changed by subclasses I figure it is very hard to fix this.

However what could be improved is the duplication:
- `with_comms` is defined at least twice and very similarly
- Unclear differences between usage of `self.world_size` and `$WORLD_SIZE`, maybe use either one?
- After the previous point maybe a (new) separate decorator `require_gpus_foreach_rank` could be used to move this out of `with_comms`
- In all those cases we can skip early (`unittest.skip`) if there are no GPUs available at all. Might already workaround a for common case (CPU only machine)